rust/actyx/ax-aql/src/language/aql.pest

```diff
@@ -7,23 +7,68 @@ main_timestamp = @{ SOI ~ isodate ~ EOI }
 
 // pragmas
 pragma = ${ "PRAGMA" ~ WHITE_SPACE+ ~ feature_word ~ ( !NEWLINE ~ WHITE_SPACE )* ~ ( pragma_multi | pragma_single ) }
-pragma_multi = _{ NEWLINE ~ pragma_multi_text ~ NEWLINE ~ "ENDPRAGMA" ~ NEWLINE }
-pragma_multi_text = ${ ( !( NEWLINE ~ "ENDPRAGMA" ~ NEWLINE ) ~ ANY )* }
+pragma_multi = _{ NEWLINE ~ pragma_multi_text ~ NEWLINE ~ pragma_end ~ NEWLINE }
+pragma_multi_text = ${ ( !( NEWLINE ~ pragma_end ~ NEWLINE ) ~ ANY )* }
+pragma_end = { "ENDPRAGMA" }
 pragma_single = _{ ":=" ~ ( !NEWLINE ~ WHITE_SPACE )* ~ pragma_single_text ~ NEWLINE }
 pragma_single_text = ${ ( !NEWLINE ~ ANY )* }
 
+features = { "FEATURES(" ~ feature_word* ~ ")" }
+feature_word = @{ ( ASCII_ALPHANUMERIC | "ø" )+ }
+
+// types
+type = { type_atom ~ type_post* ~ ( type_infix ~ type_atom ~ type_post* )* }
```

Comment:
I just realized that there is no `undefined` type which, in TS/JS, allows optional fields.
It is useful when event definition changes in a non-breaking manner so that previously persisted events can still match and work for a particular workflow.
cc: @jmg-duarte

Reply:
There is NULL — I think `undefined` is a huge mistake and don’t plan on repeating it. Optional fields should be solved differently, for example using properly defined event schema migrations. Or using default values in the type declaration.